services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/AbstractPushMonitor.java

```diff
@@ -910,8 +927,136 @@ public void onRoutingDataDeleted(String kafkaTopic) {
+    }
+  }
+
+  /**
+   * For hybrid stores we use {@link com.linkedin.venice.writer.LeaderCompleteState} to prevent follower replicas from
+   * reporting COMPLETED prematurely. The same safeguard is needed for view topic consumers. Due to the obfuscated
+   * partition mapping between version topic leaders and view topic partitions we can only send LEADER_COMPLETED once
+   * all version topic partition leaders are completed. This is equivalent to at least one completed replica for all
+   * partitions.
+   */
+  protected void checkWhetherToStartLeaderCompletedProcedures(OfflinePushStatus offlinePushStatus) {
+    if (topicToLeaderCompleteTimestampMap.containsKey(offlinePushStatus.getKafkaTopic())) {
+      // We've sent the LeaderCompleteState for all view topics of this version topic already. Duplicate heartbeats are
+      // harmless, so we are only tracking it in memory (not across restarts) to avoid heartbeat spam
+      return;
+    }
+    Store store = getStoreOrThrow(offlinePushStatus);
+    Version version = getVersionOrThrow(store, offlinePushStatus.getKafkaTopic());
+    Map<String, ViewConfig> viewConfigMap = version.getViewConfigs();
+    if (!version.isHybrid() || viewConfigMap == null || viewConfigMap.isEmpty()) {
+      return;
+    }
+    if (offlinePushStatus.isLeaderCompletedInEveryPartition()) {
+      // broadcast heartbeat with LeaderCompleteState to view topic partitions
+      long heartbeatTimestamp = System.currentTimeMillis();
+      for (ViewConfig rawViewConfig: viewConfigMap.values()) {
+        if (MaterializedView.class.getCanonicalName().equals(rawViewConfig.getViewClassName())) {
+          VeniceView veniceView = ViewUtils.getVeniceView(
+              rawViewConfig.getViewClassName(),
+              new Properties(),
+              store.getName(),
+              rawViewConfig.getViewParameters());
+          MaterializedView materializedView = (MaterializedView) veniceView;
+          for (String materializedViewTopicName: materializedView.getTopicNamesAndConfigsForVersion(version.getNumber())
+              .keySet()) {
+            VeniceWriterOptions.Builder vwOptionsBuilder =
+                new VeniceWriterOptions.Builder(materializedViewTopicName).setUseKafkaKeySerializer(true)
+                    .setPartitionCount(materializedView.getViewPartitionCount());
+            try (VeniceWriter veniceWriter = veniceWriterFactory.createVeniceWriter(vwOptionsBuilder.build())) {
+              for (int p = 0; p < materializedView.getViewPartitionCount(); p++) {
+                veniceWriter.sendHeartbeat(
+                    materializedViewTopicName,
+                    p,
+                    null,
+                    VeniceWriter.DEFAULT_LEADER_METADATA_WRAPPER,
+                    true,
+                    LeaderCompleteState.getLeaderCompleteState(true),
+                    heartbeatTimestamp);
+              }
```

Comment:
I realized that there is flaw with the agreed approach as the HB message with `LeaderCompleteState` can become stale, which will be ignored by view consumers...
It it is fine to leave this state for now and we need to think about it more about how to make it work in all cases..
Maybe we need to go back to the alternative to send HB from some leader replicas of version topics to constantly send out HB messages.
We don't need to make the code change right now as the PR is already too big.